pageserver/src/tenant/timeline/init.rs

```diff
@@ -0,0 +1,211 @@
+        .into_iter()
+        .flatten()
+        .map(|(name, metadata)| (name, LayerFileMetadata::from(metadata)))
+        .for_each(|(name, metadata)| {
+            if let Some(existing) = discovered.get_mut(name) {
+                existing.1 = Some(metadata);
+            } else {
+                discovered.insert(name.to_owned(), (None, Some(metadata)));
+            }
+        });
+
+    // must not use the index_part presence here as a proxy for remote timeline client. it might be
+    // that we have local file(s), but the index upload was stopped with failpoint. See #5075 and
+    // search for related TODO; this will need to change.
+
+    discovered
+        .into_iter()
+        .map(|(name, (local, remote))| {
+            let is_in_future = match &name {
+                Image(file_name) if file_name.lsn > disk_consistent_lsn => true,
+                Delta(file_name) if file_name.lsn_range.end > disk_consistent_lsn + 1 => true,
+                _ => false,
+            };
+
+            let decision = if is_in_future {
+                Err(FutureLayer)
+            } else {
+                Ok(match (local, remote) {
+                    (Some(local), Some(remote)) if local != remote => UseRemote { local, remote },
+                    (Some(x), Some(_)) => UseLocal(x),
+                    (None, Some(x)) => Evicted(x),
+                    (Some(x), None) => NeedsUpload(x),
+                    (None, None) => {
+                        unreachable!("there must not be any non-local non-remote files")
+                    }
+                })
+            };
+
+            (name, decision)
+        })
+        .collect::<Vec<_>>()
+}
+
+pub(super) fn cleanup(path: &Path, kind: &str) -> anyhow::Result<()> {
+    let file_name = path.file_name().expect("must be file path");
+    tracing::debug!(kind, ?file_name, "cleaning up");
+    std::fs::remove_file(path)
+        .with_context(|| format!("failed to remove {kind} at {}", path.display()))
+}
+
+pub(super) fn cleanup_local_file_for_remote(
+    path: &Path,
+    local: &LayerFileMetadata,
+    remote: &LayerFileMetadata,
+) -> anyhow::Result<()> {
+    let local_size = local.file_size();
+    let remote_size = remote.file_size();
+
+    let file_name = path.file_name().expect("must be file path");
+    tracing::warn!("removing local file {file_name:?} because it has unexpected length {local_size}; length in remote index is {remote_size}");
+    if let Err(err) = crate::tenant::timeline::rename_to_backup(path) {
+        assert!(
+            path.exists(),
+            "we would leave the local_layer without a file if this does not hold: {}",
+            path.display()
+        );
+        Err(err)
+    } else {
+        Ok(())
+    }
+}
+
+pub(super) fn cleanup_future_layer(
+    path: &Path,
+    name: LayerFileName,
+    disk_consistent_lsn: Lsn,
+) -> anyhow::Result<()> {
+    use LayerFileName::*;
+    let kind = match name {
+        Delta(_) => "delta",
```

Comment:
I don't really find this important. This is the only occurence in the codebase.

Reply:
A function would separate concerns, but shrug. I don't think it is obligatory to do this right now, we can also do it once there is an actual need somewhere else in the codebase.

Reply:
Thinking about this more, even here it is extra, because the layer file name is unambigious. However I was initially thinking of not doing a string change. Checked that this truly is the only place in the codebase.